pkg/csi/fuse/nodeserver.go

```diff
@@ -155,13 +163,138 @@ func (ns *nodeServer) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpu
 }
 
 func (ns *nodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstageVolumeRequest) (*csi.NodeUnstageVolumeResponse, error) {
+	// 1. get fuse label according to pv namespace and name
+	namespacedName := strings.Split(req.GetVolumeId(), "-")
+	fuseLabelKey := common.LabelAnnotationFusePrefix + namespacedName[0] + "-" + namespacedName[1]
+
+	// 2. check if the path is mounted
+	inUse, err := checkMountInUse(req.GetVolumeId())
+	if err != nil {
+		return nil, errors.Wrap(err, "NodeUnstageVolume: can't check mount in use")
+	}
+	if inUse {
+		return nil, fmt.Errorf("NodeUnstageVolume: can't stop fuse cause it's in use")
+	}
+
+	// 3. remove label on node.
+	// Once the label is removed, fuse pod on corresponding node will be terminated
+	// since node selector in the fuse daemonSet no longer matches.
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		node, err := kubeclient.GetNode(ns.client, ns.nodeId)
+		if err != nil {
+			return err
+		}
+
+		nodeToUpdate := node.DeepCopy()
+		delete(nodeToUpdate.Labels, fuseLabelKey)
+		if !reflect.DeepEqual(node, nodeToUpdate) {
+			return ns.client.Update(context.TODO(), nodeToUpdate)
+		} else {
+			klog.Infof("Do nothing because no label %s on node %s", fuseLabelKey, ns.nodeId)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("NodeUnstageVolume: can't remove label %s on node %s", fuseLabelKey, ns.nodeId))
+	}
 	return &csi.NodeUnstageVolumeResponse{}, nil
 }
 
 func (ns *nodeServer) NodeStageVolume(ctx context.Context, req *csi.NodeStageVolumeRequest) (*csi.NodeStageVolumeResponse, error) {
+	// 1. get fuse label according to pv namespace and name
+	namespacedName := strings.Split(req.GetVolumeId(), "-")
+	fuseLabelKey := common.LabelAnnotationFusePrefix + namespacedName[0] + "-" + namespacedName[1]
+
+	// 2. Label node
+	err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		node, err := kubeclient.GetNode(ns.client, ns.nodeId)
+		if err != nil {
+			return err
+		}
+
+		nodeToLabel := node.DeepCopy()
+		nodeToLabel.Labels[fuseLabelKey] = "true"
+		if !reflect.DeepEqual(node, nodeToLabel) {
+			return ns.client.Update(context.TODO(), nodeToLabel)
```

Comment:
Could we use patch to instead of update? It can avoid retrying.

Reply:
Done.